elasticsearch/client.go

```diff
@@ -0,0 +1,133 @@
+package elasticsearch
+
+import (
+	"bytes"
+	"context"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+
+	dphttp "github.com/ONSdigital/dp-net/http"
+	"github.com/ONSdigital/log.go/log"
+	awsauth "github.com/smartystreets/go-aws-auth"
+)
+
+// ServiceName elasticsearch
+const ServiceName = "elasticsearch"
+
+// Client is an ElasticSearch client containing an HTTP client to contact the elasticsearch API.
+type Client struct {
+	httpCli      dphttp.Clienter
+	url          string
+	serviceName  string
+	signRequests bool
+	indexes      []string
+}
+
+// NewClient returns a new initialised elasticsearch client with the default dp-net/http client
+func NewClient(url string, signRequests bool, maxRetries int, indexes ...string) *Client {
+	httpClient := dphttp.NewClient()
+	httpClient.SetMaxRetries(maxRetries)
+	return NewClientWithHTTPClient(url, signRequests, httpClient, indexes...)
+}
+
+// NewClientWithHTTPClient returns a new initialised elasticsearch client with the provided HTTP client
+func NewClientWithHTTPClient(url string, signRequests bool, httpCli dphttp.Clienter, indexes ...string) *Client {
+	return &Client{
+		httpCli:      httpCli,
+		url:          url,
+		serviceName:  ServiceName,
+		signRequests: signRequests,
+		indexes:      indexes,
+	}
+}
+
+//CreateIndex creates an index in elasticsearch
+func (cli *Client) CreateIndex(ctx context.Context, path string, content []byte) (int, error) {
```

Comment:
Currently a mapping is passed into the create index call in the search builder, but I am not too sure what it does: https://github.com/ONSdigital/dp-search-builder/blob/develop/elasticsearch/api.go#L48

Reply:
I changed the final parameter to indexSettings of type byte, because from the documentation https://www.elastic.co/guide/en/elasticsearch/reference/6.8/indices-create-index.html it seems that index settings can be passed in the body of the call to elasticsearch, and I think this is what the mappings in the search builder represents. I thought to use a variadic function because if no index settings are required nothing has to be passed in?